doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,258 @@
+
+## `link` (import-only): invoked before linking
+
+This is invoked after `load` but prior to Node.js passing the final result to V8 for linking, so it can be composed with `resolve` and `load` if necessary. This needs to work with experimental vm modules for passing module instances around.
+
+```js
+/**
+ * @typedef {{
+ *  source: string | Buffer
+ * }} ModuleLinkContext
+ */
+/**
+ * @typedef {{
+*   module: vm.Module
+* }} ModuleLinkResult
+*/
+/**
+ * @param {string} url
+ * @param {ModuleLinkContext} context
+ * @param {(context: ModuleLinkContext) => {ModuleLinkResult}} nextLink
+ * @returns {ModuleLinkResult}
+ */
+function link(url, context, nextLink) {  // Mocking import-in-the-middle
+  const { module: originalModule } = nextLink(url, context);
+  assert.strictEqual(module.status, 'linked');  // Original module is linked at this point
+  let source = `import * as original from 'original';`;
+  source += `import { userCallback, name, basedir } from 'util'`;
+  source += `const exported = {}`;
+  for (const key of originalModule.namespace) {
+    source += `let $${key} = original.${key};`;
+    source += `export { $${key} as ${key} }`;
+    source += `Object.defineProperty(exported, '${key}', { get() { return $${key}; }, set (value) { $${key} = value; }});`;
+  }
+  source += `userCallback(exported, name, basedir);`;
+  const m = vm.SourceTextModule(source);
+  m.linkSync((specifier) => {  // This is not yet implemented but should be trivial to implement.
+    if (specifier === 'original') return originalModule;
+    // Contains a synthetic module with userCallback, name & basedir computed from url
+    if (specifier === 'util') return util;
+  });
+  return { module: m };
+}
+```
+
+### Alternative design: `link` that encompasses `resolve` and `load`
+
+An alternative design would be, instead of invoking a hook after `load`, make it wrap around `resolve` and `load` steps. The `link` hooks would take `specifier` and return `vm.Module` instances. If `nextLink()` is invoked, the next `resolve` and `load` (e.g. the default ones) will be invoked inside that. If `nextLink()` is not invoked, the default `resolve` and `load` will be skipped.
+
+```js
+function link(specifier, context, nextImport) {  // Mocking import-in-the-middle
+  const { module: originalModule } = nextLink(specifier, context);
+  assert.strictEqual(module.status, 'linked');  // Original module is linked at this point
+  let source = '...';
+  const m = vm.SourceTextModule(source);
+  m.linkSync((specifier) => {
+    // ...
+  });
+  return { module: m };
+}
+```
+
+## Other use cases that may need some thought
+
+1. Hot module replacement
+2. Source map support (e.g. see [babel-register](https://github.com/babel/babel/blob/07bd0003cbdaa8525279c6dfa84e435471eb5797/packages/babel-register/src/hook.js#L38))
+3. Virtual file system in packaged apps (single executable applications).
```

Comment:
```suggestion
3. Virtual file system in packaged apps (single executable applications).
4. Asynchronous work inside hooks will be delegated to an off-thread solution via a "deasync" pattern using `Atomic`. Providing this kind of plumbing is outside of the scope of Node.js core.
```

Do we want to have such an utility in core or should we point it to an ecosystem module?

Reply:
It seems better to just let some ecosystem module handle it. When that matures it might be interesting to bring it in core but developing it in core from the start when it can be done in the user land directly isn't a great pattern IMO.